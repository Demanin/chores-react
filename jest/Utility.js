import crypto from 'crypto';

const getRandomString = (length = 20) => {
  return crypto.randomBytes(length).toString('hex');
};

const getRandomNumber = (min = 0, max = 9999999) => {
  min = Math.ceil(min);
  max = Math.floor(max);

  return Math.floor(Math.random() * (max - min)) + min;
};

const getRandomBool = () => {
  return !!getRandomNumber(0, 1);
};

export default {
  getRandomString,
  getRandomNumber,
  getRandomBool,
};
