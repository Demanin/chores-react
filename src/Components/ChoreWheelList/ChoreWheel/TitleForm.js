import React, { Component } from 'react';
import { FormGroup, FormControl } from 'react-bootstrap';

class TitleForm extends Component
{
  constructor(props) {
    super(props);

    this.handleChange  = this.handleChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);

    this.state = {
      value: props.title
    };
  }

  handleChange(event) {
    this.setState({ value: event.target.value });
  }

  handleKeyDown(event) {
    if (event.keyCode === 13) {
      this.props.onSave(this.state.value);
    }
  }

  render() {
    return (
      <FormGroup
        controlId={"title-editor-" + this.props.id}
      >
        <FormControl
          autoFocus
          value={this.state.value}
          onChange={this.handleChange}
          onBlur={() => this.props.onSave(this.state.value)}
          onKeyDown={this.handleKeyDown}
        />
      </FormGroup>
    );
  }
}

export default TitleForm;
