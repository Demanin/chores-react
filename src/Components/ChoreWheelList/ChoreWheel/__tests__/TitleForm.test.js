import React from 'react';
import TitleForm from 'src/Components/ChoreWheelList/ChoreWheel/TitleForm';
import Utility from 'jest/Utility';
import { shallow } from 'enzyme';

describe('TitleForm', () => {
  let onSave;

  beforeEach(() => {
    onSave = jest.fn();
  });

  test('empty form snapshot', () => {
    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ onSave }}
    />);

    expect(fixture).toMatchSnapshot();
  });

  test('FormGroup controlId', () => {
    const id = Utility.getRandomNumber();

    const fixture = shallow(<TitleForm
      choreWheel={{ id }}
      {...{ onSave }}
    />);

    expect(fixture.prop('controlId')).toEqual('title-editor-' + id);
  });

  test('FormControl value starts with title', () => {
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title, onSave }}
    />);

    expect(fixture.find('FormControl').prop('value')).toEqual(title);
  });

  test('blur immediately calls onSave with original title', () => {
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title, onSave }}
    />);

    fixture.find('FormControl').simulate('blur');

    expect(onSave).toHaveBeenCalledWith(title);
  });

  test('pressing enter immediately calls onSave with original title', () => {
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title, onSave }}
    />);

    fixture.find('FormControl').simulate('keydown', { keyCode: 13 });

    expect(onSave).toHaveBeenCalledWith(title);
  });

  test('non-enter key does not save', () => {
    const keyCode = Utility.getRandomNumber(14, 100);
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title, onSave }}
    />);

    fixture.find('FormControl').simulate('keydown', { keyCode });

    expect(onSave).not.toHaveBeenCalled();
  });

  test('blurring after changing title saves new title', () => {
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title: Utility.getRandomString(), onSave }}
    />);

    fixture.find('FormControl').simulate('change', { target: { value: title } });
    fixture.find('FormControl').simulate('blur');

    expect(onSave).toHaveBeenCalledWith(title);
  });

  test('pressing enter after changing title saves new title', () => {
    const title = Utility.getRandomString();

    const fixture = shallow(<TitleForm
      choreWheel={{}}
      {...{ title: Utility.getRandomString(), onSave }}
    />);

    fixture.find('FormControl').simulate('change', { target: { value: title } });
    fixture.find('FormControl').simulate('keydown', { keyCode: 13 });

    expect(onSave).toHaveBeenCalledWith(title);
  });
});
