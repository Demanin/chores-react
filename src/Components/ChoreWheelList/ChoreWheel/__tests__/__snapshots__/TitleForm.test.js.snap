// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleForm empty form snapshot 1`] = `ShallowWrapper {}`;
